contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,193 @@
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
+kubectl (and client-go) should be able to receive credentials by executing a
+command and reading that command's stdout.
+
+In fact, kubectl already does this today. The GCP plugin can already be configured
+to [call a command](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.8.5/plugin/pkg/client/auth/gcp/gcp.go#L228-L240)
+other than `gcloud`.
+
+## Kubeconfig changes
+
+The current `AuthProviderConfig` uses `map[string]string` for configuration, which
+makes it hard to express things like a list of arguments or list key/value environment
+variables. As such, `AuthProviderConfig` should add another field which expresses the
+`exec` config. This has the benefit of a more natural structure, but the trade-off of
+not being compatible with the existing `kubectl config set-credentials` implementation.
+
+```go
+const (
+    AuthProviderExecKindBearerToken = "bearer-token"
+    AuthProviderExecKindClientCert  = "tls-certificate"
+)
+
+type AuthProviderConfig struct {
+    Name   string            `json:"name"`
+    Config map[string]string `json:"config"`
+
+    // Exec should only be provided if Name and Config aren't.
+    Exec ExecAuthProviderConfig `json:"exec"`
+}
+
+type ExecAuthProviderConfig struct {
+    Command string   `json:"command"`
+    Args    []string `json:"args"`
+    // Env defines additional environment variables to expose to the process. These
+    // are unioned with the host's environment, with variable values lists here having
+    // precedence.
+    Env []ExecEnvVar `json:"env"`
+    // Kind determines the kind of credentials the command will output to stdout.
+    // This can be "bearer-token" or "tls-certificate".
+    Kind string `json:"kind"`
+}
+
+type ExecEnvVar struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+
+    // TODO: Load env vars from files or from other envs?
+}
+```
+
+This would allow a user block of a kubeconfig to declare the following:
+
+```yaml
+users:
+- name: mmosley
+  user:
+    auth-provider:
+      exec:
+        kind: bearer-token
+        command: /bin/kubectl-login
+        args: ["hello", "world"]
+        env:
+        - name: "VERSION"
+          value: "v1.0.1"
+```
+
+The AWS authenticator above example becomes:
+
+```
+users:
+- name: kubernetes-admin
+  user:
+    auth-provider:
+      exec:
+        kind: bearer-token
+        command: heptio-authenticator-aws
+        # CLUSTER_ID and ROLE_ARN should be replaced with actual desired values.
+        args: ["token", "-i", "(CLUSTER_ID)", "-r", "(ROLE_ARN)"]
+```
+
+## TLS client certificate support
```

Comment:
This seems like a related but separate change from the executable plugin. It is not necessary to break it out into its own proposal but I wanted to note that one does not block the other from moving forward.

Reply:
Thanks, added a comment.